internal/sourceanalysis/rust.go

```diff
@@ -0,0 +1,283 @@
+	if err != nil {
+		return "", fmt.Errorf(".rlib file '%s' is not valid ar archive: %w", rlibPath, err)
+	}
+	for {
+		header, err := reader.Next()
+		if err != nil {
+			log.Fatalf("%v", err)
+		}
+		if header.Name == "//" { // "//" is used in GNU ar format as a store for long file names
+			fileBuf := bytes.Buffer{}
+			// Ignore the error here as it's likely
+			_, err = io.Copy(&fileBuf, reader)
+			if err != nil {
+				return "", fmt.Errorf("failed to read // store in ar archive: %w", err)
+			}
+			// There should only be one file (since we set codegen-units=1)
+			if !strings.HasSuffix(fileBuf.String(), RustLibExtension) {
+				// TODO: Verify this, and return an error here instead.
+				log.Printf("rlib archive contents were unexpected: %s\n", fileBuf.String())
+			}
+		}
+		// /0 indicates the first file mentioned in the "//" store
+		if header.Name == "/0" || strings.HasSuffix(header.Name, RustLibExtension) {
+			break
+		}
+	}
+	_, err = io.Copy(file, reader)
+	if err != nil {
+		return "", fmt.Errorf("failed to write to temporary file '%s': %w", file.Name(), err)
+	}
+
+	filePath := file.Name()
+	err = file.Close()
+	if err != nil {
+		return "", fmt.Errorf("failed to close the temporary file '%s': %w", file.Name(), err)
+	}
+
+	return filePath, nil
+}
+
+func rustBuildSource(r reporter.Reporter, source models.SourceInfo) ([]string, error) {
+	projectBaseDir := filepath.Dir(source.Path)
+
+	cmd := exec.Command("cargo", "build", "--all-targets", "--release")
+	cmd.Env = append(cmd.Environ(), RustFlagsEnv)
+	cmd.Dir = projectBaseDir
+	if errors.Is(cmd.Err, exec.ErrDot) {
+		cmd.Err = nil
+	}
+
+	stdoutBuffer := bytes.Buffer{}
+	stderrBuffer := bytes.Buffer{}
+	cmd.Stdout = &stdoutBuffer
+	cmd.Stderr = &stderrBuffer
+
+	r.PrintText("Begin building rust/cargo project\n")
+
+	if err := cmd.Run(); err != nil {
+		r.PrintError(fmt.Sprintf("cargo stdout:\n%s", stdoutBuffer.String()))
+		r.PrintError(fmt.Sprintf("cargo stderr:\n%s", stderrBuffer.String()))
+
+		return nil, fmt.Errorf("failed to run cargo build: %w", err)
+	}
+
+	outputDir := filepath.Join(projectBaseDir, "target", "debug")
+	entries, err := os.ReadDir(outputDir)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read \"%s\" dir: %w", outputDir, err)
+	}
+
+	resultBinaryPaths := []string{}
+	for _, de := range entries {
+		// We only want .d files, which is generated for each output binary from cargo
+		// These files contains a string to the full path of output binary/library file.
+		// This is a reasonably reliable way to identify the output in a cross platform way.
+		if de.IsDir() || !strings.HasSuffix(de.Name(), ".d") {
+			continue
+		}
+
+		file, err := os.ReadFile(filepath.Join(outputDir, de.Name()))
+		if err != nil {
+			return nil, fmt.Errorf("failed to read \"%s\": %w", filepath.Join(outputDir, de.Name()), err)
+		}
+
+		fileSplit := strings.Split(string(file), ": ")
+		if len(fileSplit) != 2 {
+			// TODO: this can probably be fixed with more effort
```

Comment:
IS this something strings.SplitN  could help with?

Reply:
Not really, the `.d` format is very simple, where there's two paths separated by a `:<space>`. The problem happens when the path itself contains a `:<space>`. It doesn't seem to get escaped by rust compiler. It's going to be pretty rare though, so I'm not sure if it's worth coming up with a better solution.